docs/device-api.md

```diff
@@ -0,0 +1,106 @@
+# Device-API Functionality
+
+The Device-API following functionality matrix indicates which devices are
+required for an API call to be usable.
+
+**O** (Optional): The device (column) **is not required** for a Firecracker
+microVM to run. If the device (column) is omitted from a kernel build the API
+element (row) is **TBA**.
+
+**R** (Required): The device (column) **is required** for a Firecracker microVM
+to run. If the device (column) is omitted from a kernel build the API
+element (row) response is **TBA**.
+
+## API Endpoints
+
+| Endpoint                  | keyboard | serial console | virtio-block | virtio-net | virtio-vsock |
+|---------------------------|:--------:|:--------------:|:------------:|:----------:|:------------:|
+| `boot-source`             |    O     |    O           |   O          |   O        |   O          |
+| `drives/{id}`             |    O     |    O           | **R**        |   O        |   O          |
+| `logger`                  |    O     |    O           |   O          |   O        |   O          |
+| `machine-config`          |    O     |    O           |   O          |   O        |   O          |
+| `metrics`                 |    O     |    O           |   O          |   O        |   O          |
+| `mmds`                    |    O     |    O           |   O          | **R**      |   O          |
+| `mmds/config`             |    O     |    O           |   O          | **R**      |   O          |
+| `network-interfaces/{id}` |    O     |    O           |   O          | **R**      |   O          |
+| `snapshot/create`         |    O     |    O           |   O          |   O        |   O          |
+| `snapshot/load`           |    O     |    O           |   O          |   O        |   O          |
+| `vm`                      |    O     |    O           |   O          |   O        |   O          |
+| `vsock`                   |    O     |    O           |   O          |   O        | **R**        |
+
+## Input Schema
+
+| Schema                    | Property              | keyboard | serial console | virtio-block | virtio-net | virtio-vsock |
+|---------------------------|-----------------------|:--------:|:--------------:|:------------:|:----------:|:------------:|
+| `BootSource`              | boot_args             |   O      |   O            |   O          |   O        |   O          |
```

Comment:
Utterly confused by that description.

Reply:
> IMO that aid is best given by being explicit and not having them decode some summary/abbreviation conventions.

I see. My concern here is that we are required to keep APIs, input and outputs schemas duplicated in both swagger file and Device-API matrix. Users will eventually want to checkout the swagger file (and understand the semantics of this file) if they are interested in the source of truth for APIs and schema definitions. When users want to understand what devices are required for specific APIs endpoints or schemas, they need to look it up in the Device-API matrix. Having an abbreviation convention means that some API endpoint, input or output schema will fall under the wildcard, which means that they will not explicitly appear in the document, which might cause confusion for end users. Adding a backing definition in the document for the wildcard will address this issue.

In the end, I would like to say that I do not have a strong preference for my suggestion. It was just a callout for getting more comments and see if it is something which you or others would prefer instead of the long tables form. Feel free to choose the option you think it is better for end users, I am OK with both.

Reply:
> My concern here is that we are required to keep APIs, input and outputs schemas duplicated in both swagger file and Device-API matrix.

Fully agree and share that concern.  However we are documenting the connection between two distinct systems: Swagger API and Firecracker uVM, so it seems unavoidable that some hand maintained document is required.
IIUC, Firecracker's focus means I'd expect its device model and even API to be very slowly changing, if at all once v1 is released. so changes here should be infrequent.

In case it turns out there are frequent of changes:
We might (eventually) be able to script this table from some Firecracker JSON master definition file (is there such a thing?) and the Swagger API definition?  Updating this document would be part of the release process.
But that is not possible now, and may require adding linking data to their Swagger or JSON master file.

I'm having trouble understanding the conventions you suggest and how they would work. 
Maybe they can be made as a PR once this lands?

Reply:
We do not have frequent changes in API area, but if we do, we need to update the release process to sync the Device-API matrix. We can further improve this documentation in following PRs, if needed. Will resolve this comment for now.